services/bootnode/src/main/java/teku/pegasys/teku/services/bootnode/BootnodeService.java

```diff
@@ -0,0 +1,119 @@
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.networking.eth2.P2PConfig;
+import tech.pegasys.teku.networking.p2p.discovery.DiscoveryService;
+import tech.pegasys.teku.networking.p2p.discovery.discv5.DiscV5Service;
+import tech.pegasys.teku.networking.p2p.libp2p.LibP2PNetwork.PrivateKeyProvider;
+import tech.pegasys.teku.networking.p2p.libp2p.LibP2PPrivateKeyLoader;
+import tech.pegasys.teku.networking.p2p.network.config.NetworkConfig;
+import tech.pegasys.teku.networks.Eth2NetworkConfiguration;
+import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.service.serviceutils.ServiceConfig;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionsSupplier;
+import tech.pegasys.teku.storage.store.FileKeyValueStore;
+import tech.pegasys.teku.storage.store.KeyValueStore;
+
+public class BootnodeService extends Service {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private static final String KEY_VALUE_STORE_SUBDIRECTORY = "kvstore";
+  private static final int PRINT_KNOWN_NODES_COUNT_INTERVAL_IN_SECONDS = 5;
+
+  private final DiscoveryService discoveryService;
+  private final AsyncRunner asyncRunner;
+
+  public BootnodeService(
+      final ServiceConfig serviceConfig,
+      final NetworkConfig networkConfig,
+      final P2PConfig p2PConfig,
+      final Eth2NetworkConfiguration eth2NetworkConfiguration) {
+    final MetricsSystem metricsSystem = serviceConfig.getMetricsSystem();
+    checkNotNull(metricsSystem);
+
+    asyncRunner =
+        serviceConfig.createAsyncRunner(
+            "bootnode",
+            eth2NetworkConfiguration.getAsyncP2pMaxThreads(),
+            eth2NetworkConfiguration.getAsyncP2pMaxQueue());
+    final KeyValueStore<String, Bytes> kvStore =
+        new FileKeyValueStore(
+            serviceConfig
+                .getDataDirLayout()
+                .getBootnodeDataDirectory()
+                .resolve(KEY_VALUE_STORE_SUBDIRECTORY));
+
+    final SchemaDefinitionsSupplier currentSchemaDefinitionsSupplier =
+        () -> eth2NetworkConfiguration.getSpec().getGenesisSchemaDefinitions();
+    checkNotNull(currentSchemaDefinitionsSupplier);
+
+    final PrivateKeyProvider privateKeyProvider =
+        new LibP2PPrivateKeyLoader(kvStore, networkConfig.getPrivateKeySource());
+
+    discoveryService =
+        new DiscV5Service(
+            metricsSystem,
+            asyncRunner,
+            p2PConfig.getDiscoveryConfig(),
+            networkConfig,
+            kvStore,
+            Bytes.wrap(privateKeyProvider.get().raw()),
+            currentSchemaDefinitionsSupplier,
+            DiscV5Service.createDefaultDiscoverySystemBuilder(),
+            DiscV5Service.DEFAULT_NODE_RECORD_CONVERTER);
+
+    asyncRunner.runWithFixedDelay(
+        () -> LOG.info("Total peers known: {}", discoveryService.streamKnownPeers().count()),
```

Comment:
is this just so that we have data as we go?

Reply:
Yeah mostly a way of knowing we are finding nodes etc.